panel/util.py

```diff
@@ -290,6 +296,91 @@ def add_to_doc(obj, doc, hold=False):
         doc.hold()
 
 
+def record_events(doc):
+    events = list(doc._held_events)
+    if not events:
+        return None
+    msg = Protocol("1.0").create("PATCH-DOC", events, use_buffers=False)
+    return {'header': msg.header_json, 'metadata': msg.metadata_json,
+            'content': msg.content_json}
+
+
+def embed_state(panel, model, doc, max_states=1000):
+    """
+    Embeds the state of the application on a State model which allows
+    exporting a static version of an app. This works by finding all
+    widgets with a predefined set of options and evaluating the cross
+    product of the widget values and recording the resulting events to
+    be replayed when exported. The state is recorded on a State model
+    which is attached as an additional root on the Document.
+
+    Parameters
+    ----------
+    panel: panel.viewable.Reactive
+      The Reactive component being exported
+    model: bokeh.model.Model
+      The bokeh model being exported
+    doc: bokeh.document.Document
+      The bokeh Document being exported
+    max_states: int
+      The maximum number of states to export
+    """
+    from .models.state import State
+    from .widgets import Widget
+
+    target = model.ref['id']
+    model.tags.append('embedded')
+    widgets = [w for w in panel.select(Widget) if 'options' in w.params()]
+
+    add_to_doc(model, doc, True)
+
+    keys = []
+    values = []
+    for w in widgets:
+        w_model = w._models[target].select_one({'type': w._widget_type})
+        js_callback = CustomJS(code="""
+          var receiver = new Bokeh.protocol.Receiver()
+          state = cb_obj.document.roots()[1]
+          msg = state.get_state(cb_obj)
+          receiver.consume(msg.header)
+          receiver.consume(msg.metadata)
+          receiver.consume(msg.content)
+          if (receiver.message)
+            cb_obj.document.apply_json_patch(receiver.message.content)
+        """)
+        w_model.js_on_change('value', js_callback)
+        if isinstance(w.options, list):
+            values.append((w, w_model, w.options))
+        else:
+            values.append((w, w_model, list(w.options.values())))
+    doc._held_events = []
+
+    state_dict = defaultdict(dict)
+    restore = [w.value for w, _, _ in values]
+    init_vals = [m.value for _, m, _ in values]
+    cross_product = list(product(*[vals[::-1] for _, _, vals in values]))
+
+    if len(cross_product) > max_states:
+        raise RuntimeError('The cross product of different application '
+                           'states is too large to explore (N=%d), either reduce '
+                           'the number of options on the widgets or increase '
+                           'the max_states specified on static export.' %
+                           len(cross_product))
```

Comment:
Is there any hope that exceeding the max_states would be a warning rather than an error, such that instead of aborting Panel decimates the widget options included (at least the numeric ones) until max_states is satisfied?